builder/build_validate.go

```diff
@@ -0,0 +1,927 @@
+
+	for p := range toPkgs {
+		if fromPkgs[p] != false {
+			// Match
+			if isPkgMod(p, bundleDiff) {
+				modList = append(modList, p)
+			}
+		} else {
+			// Added pkg
+			addList = append(addList, p)
+		}
+	}
+
+	for p := range fromPkgs {
+		if toPkgs[p] == false {
+			// Deleted pkg
+			delList = append(delList, p)
+		}
+	}
+	return diffLists{modList: modList, addList: addList, delList: delList}
+}
+
+func getManFileDiffLists(fromFiles, toFiles map[string]*swupd.File) diffLists {
+	addList := []string{}
+	delList := []string{}
+	modList := []string{}
+
+	for f := range toFiles {
+		if fromFiles[f] != nil {
+			// Match
+			if isManFileMod(fromFiles[f], toFiles[f]) {
+				modList = append(modList, f)
+			}
+		} else {
+			// Added file
+			addList = append(addList, f)
+		}
+	}
+
+	for f := range fromFiles {
+		if toFiles[f] == nil {
+			// Deleted file
+			delList = append(delList, f)
+		}
+	}
+	return diffLists{modList: modList, addList: addList, delList: delList}
+}
+
+func isFileMod(from, to *fileInfo) bool {
+	return ((from.name != to.name) ||
+		(from.hash != to.hash) ||
+		(from.modes != to.modes) ||
+		(from.links != to.links) ||
+		(from.user != to.user) ||
+		(from.group != to.group))
+}
+
+func isPkgMod(key string, bundleDiff *mcaBundleDiff) bool {
+	// When file(s) associated with a package are modified, the package is modified
+	if (bundleDiff.pkgFileCounts[key].add + bundleDiff.pkgFileCounts[key].mod + bundleDiff.pkgFileCounts[key].del) != 0 {
+		return true
+	}
+	return false
+}
+
+func isManFileMod(from, to *swupd.File) bool {
+	return ((from.Name != to.Name) ||
+		(from.Hash != to.Hash))
+}
+
+// analyzeMcaResults compares manifest file changes against package file changes.
+// When there are inconsistencies between the manifest and package file lists,
+// a slice of error strings is returned.
+func analyzeMcaResults(results *mcaDiffResults, fromInfo, toInfo map[string]*mcaBundleInfo) ([]string, error) {
+	var errorList = []string{}
+	var tmp = []string{}
+
+	if results.bundleDiff["os-core"] == nil {
+		return nil, fmt.Errorf("os-core does not have mandatory file changes")
+	}
+
+	// The os-release, version, and versiontimestamp files are required to be
+	// modified by the os-core bundle and are not included by an rpm. Confirm
+	// that these files were modified in the manifest, but omit them from the
+	// package comparison.
+	mandatoryFiles := 0
+	for _, f := range results.bundleDiff["os-core"].manFileDiffs.modList {
+		switch f {
+		case "/usr/lib/os-release":
+			mandatoryFiles++
+		case "/usr/share/clear/version":
+			mandatoryFiles++
+		case "/usr/share/clear/versionstamp":
+			mandatoryFiles++
+		default:
+			tmp = append(tmp, f)
+		}
+	}
+	results.bundleDiff["os-core"].manFileDiffs.modList = tmp
+
+	if mandatoryFiles != 3 {
+		errorList = append(errorList, "ERROR: os-core is missing mandatory modified file(s)\n")
+	}
+
+	// Compare manifest files against package files. Inconsistencies are added to the errorList.
```

Comment:
What about other files added directly to the chroot between "build bundles" and "build update"? Won't they cause trouble here?

Reply:
Inserting unique files wouldn't cause any problems for MCA because the bundle manifests and MoM would be unchanged (full manifest is not checked by MCA).

Altering files that exist in bundle manifests may be overlooked by MCA because it doesn't directly compare package content against manifest content. MCA confirms that every package file change has a corresponding manifest file change, but the content of the changes is left unchecked. So, if a specific file in a manifest is expected to change, MCA will succeed as long as that file is modified in any way.